### ISSUE TITLE
Feature suggestion: Extendable Registration Profiles

### ISSUE THREAD

**Issue by dicato** [macropin/django-registration#18]
While writing a custom backend for django-registration, I've had the requirement to extend the provided `RegistrationProfile` with additional fields to request from a user. For example, maybe you require knowing the user's `company` or `location`.

After a couple days of working, I have a very custom solution that works for my django project. It required:

* subclass `registration.models.RegistrationProfile` and `registration.models.RegistrationManager` in order to add fields to the `RegistrationProfile` and account for them in `RegistrationManager.create_inactive_user`
* subclass `registration.admin.RegistrationAdmin` in order to expose the new fields on `RegistrationProfile`. This mostly copy-and-pasted code.
* subclass `registration.forms.RegistrationForm` in order to add the additional `RegistrationProfile` fields and expose them during the registration process.
* subclass `registration.backends.default.views.RegistrationView` in order to override the `form_class` specifying my custom form as well as overriding `RegistrationView.register` in order to be able to pass the additional `RegistrationProfile` fields through to `create_inactive_user`.
* copy-and-paste the `registration.backends.default.urls` *exactly*, only overriding the imports for my custom views instead of the default backed views.
* remove `registration` from my project's `settings.py`, specifying my custom django-registration "backend application" instead, since at this point it's more of a custom django application than an overriding django-registration backend.

After doing this work, I started to think it would make sense to refactor the base views in `registration` as well as the provided backend views in order to allow the model being used to be more "pluggable". Specifically, follow the implementation of the `form_class` attribute in `registration.views.RegistrationView`, allowing the author of the custom backend to optionally specify a custom `RegistrationProfile`; a custom model.

I'm still thinking through potential side effects to backwards compatibility and project settings (for example my last bullet above mentioning the changes to my project's `settings.py`).

Is this something the project would benefit from? Does my thought process at least make sense? I'm going to spend some time working on this, but would love any feedback or ideas.

**Comment by theneel** [comment]
I am trying to figure out that, have you reached a proper solution?

**Comment by macropin** [comment]
Have you considered a [custom user model](https://docs.djangoproject.com/en/1.7/topics/auth/customizing/#specifying-a-custom-user-model) for the additional fields? This is one way, alternatively a account model with a 1:1 on the user can be used. I don't think extending the `RegistrationProfile` is the right approach. What is the benefit?

Cheers, Andrew

**Comment by Locker537** [comment]
@macropin,
Thanks for the suggestion. For my use case, I avoided a customer user model for the exact reason described in the documentation. It's usually more practical as you suggested to have a model, call it `UserProfile`, with a foreign key or one-to-one relation back to `User`.

I actually have a similar setup, but a `GroupProfile` model with a one-to-one relationship to `Group`. `GroupProfile` has additional foreign key relationships to other models as well.

During the user registration we process, we capture information related to both the `GroupProfile` and the additional models related to `GroupProfile`.

Essentially we capture a variety of data spanning three models, and don't want to create instances of those models unless we, as administrators of the service, approve the user and the provided details. I'm not sure if there is a better design for this process, but it's non-trivial to change it regardless. We might be overloading the functionality of django-registration, but it's working quite well for us nonetheless.

Does that use case scenario make more sense, even if it might be overly unique?

**Comment by LennyLip** [comment]
I think RegistrationForm must be ModelForm of the UserModel() with fields-dependent behavior... May be with several basic customizable fields, e.g. using Django setting.USERNAME_FIELD insted of UserModel().username, etc..
But it is big work for future.

**Comment by willfurnell** [comment]
@dicato Did you get this sorted in the end? I would love to have the user need to enter a 'secret' word and choose something from a dropdown during the registration process to put in the UserProfile model.

**Comment by dicato** [comment]
@willfurnell we used the above approach for awhile. When a `User` registered, we forced them to provide two additional pieces of information. We saved that information on our custom `RegistrationProfile` and then used it to pre-configure settings in our application.

Feedback from our users and a few people with expertise in user experience told us they did not like this process. We had created additional steps in our registration process and were absolutely loosing users because of it.

Instead, we now have a very "standard" registration process where the `User` provides their username, email, and a password (only once by the way) and *is immediately registered and logged in*. The additional fields we were storing on the `RegistrationProfile` have been moved to custom models that behave akin to `UserProfile` and `GroupProfile`. We actually use Django's messages to indicate to the User what settings they *must* or *should* configure.

To summarize, we choose to store the fields we care about on other models (`UserProfile`, `GroupProfile`) and decouple the User entering that information from the registration process in order to make the registration process *as easy as possible*.

----

I'm going to close this issue.

**Comment by rylanchiu** [comment]
@dicato Could you please explain more on your last approach? The registration just need username, email and password, but when do you require them to finish their profile? Thanks.

**Comment by dicato** [comment]
@rylanchiu I've tried to address your question in the issue you filed about collecting User provided images: #123
